### PR TITLE
Complete logging for Socket.* packages

### PR DESCRIPTION
**Pull request by BrennanConroy**
Part of https://github.com/aspnet/SignalR/issues/553

### CODE REVIEW

**Comment by BrennanConroy** [comment]
@davidfowl Question: Should we be using `DateTime.UtcNow`? Gets rid of any ambiguity when comparing logs, plus on my machine `DateTime.UtcNow` is over 10 times faster than `DateTime.Now`

**Comment by davidfowl** [comment]
> @davidfowl Question: Should we be using DateTime.UtcNow? Gets rid of any ambiguity when comparing logs, plus on my machine DateTime.UtcNow is over 15 times faster than DateTime.Now

I don't know if we need it at all though.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 1]
```diff
@@ -36,18 +38,18 @@ public LongPollingTransport(HttpClient httpClient, ILoggerFactory loggerFactory)
 
         public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
         {
-            _logger.LogInformation("Starting {0}", nameof(LongPollingTransport));
+            _connectionId = System.Web.HttpUtility.ParseQueryString(url.Query)["id"];
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Nope. Pass it in.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/SocketLoggerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,502 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.WebSockets;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Sockets.Internal
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/SocketLoggerExtensions.cs]: Should this namespace be Microsoft.AspNetCore.Sockets.**Client**.Internal ?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 3]
```diff
@@ -34,20 +36,20 @@ public LongPollingTransport(HttpClient httpClient, ILoggerFactory loggerFactory)
             _logger = (loggerFactory ?? NullLoggerFactory.Instance).CreateLogger<LongPollingTransport>();
         }
 
-        public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
+        public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, string connectionId)
         {
-            _logger.LogInformation("Starting {0}", nameof(LongPollingTransport));
+            _connectionId = connectionId;
+            _logger.StartTransport(_connectionId);
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: Why did you remove the transport?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs [hunk 4]
```diff
@@ -57,7 +59,7 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application)
 
         public async Task StopAsync()
         {
-            _logger.LogInformation("Transport {0} is stopping", nameof(LongPollingTransport));
+            _logger.TransportStopping(_connectionId);
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: transport?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs [hunk 5]
```diff
@@ -7,7 +7,7 @@
 
 namespace Microsoft.AspNetCore.Sockets.Internal
 {
-    public static class SocketLoggerExtensions
+    internal static class SocketLoggerExtensions
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs]: Why `internal`? (just asking)

File: test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs [hunk 6]
```diff
@@ -129,6 +129,11 @@ public EndToEndTests(ServerFixture serverFixture, ITestOutputHelper output) : ba
 
                     await closeTcs.Task.OrTimeout();
                 }
+                catch (Exception ex)
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs]: why is it useful? wouldn't the exception appear twice?

File: test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs [hunk 7]
```diff
@@ -37,7 +37,7 @@ public void DefaultHubProtocolResolverThrowsForNotSupportedProtocol(string proto
             Assert.Equal($"The protocol '{protocolName ?? "(null)"}' is not supported.", exception.Message);
         }
 
-        public static IEnumerable<object[]> HubProtocols() =>
+        public static IEnumerable<object[]> HubProtocols =>
```
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Tests/Internal/DefaultHubProtocolResolverTests.cs]: :+1:

File: src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/SocketLoggerExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+
+        private static readonly Action<ILogger, DateTime, string, Exception> _sendCanceled =
+            LoggerMessage.Define<DateTime, string>(LogLevel.Debug, 8, "{time}: Connection Id {connectionId}: Send loop canceled.");
+
+        // Category: WebSocketsTransport
+        private static readonly Action<ILogger, DateTime, string, WebSocketCloseStatus?, Exception> _webSocketClosed =
+            LoggerMessage.Define<DateTime, string, WebSocketCloseStatus?>(LogLevel.Information, 9, "{time}: Connection Id {connectionId}: Websocket closed by the server. Close status {closeStatus}.");
+
+        private static readonly Action<ILogger, DateTime, string, WebSocketMessageType, int, bool, Exception> _messageReceived =
+            LoggerMessage.Define<DateTime, string, WebSocketMessageType, int, bool>(LogLevel.Debug, 10, "{time}: Connection Id {connectionId}: Message received. Type: {messageCount}, size: {count}, EndOfMessage: {endOfMessage}.");
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/Internal/SocketLoggerExtensions.cs]: messageCount should be messageType

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs (hunk 4 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: (same in other transports)

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs (hunk 3 above)
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: It seemed silly to include the transport in the string when the category is the transport

File: test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs (hunk 6 above)
Comment (BrennanConroy) [inline test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs]: Exceptions thrown can be hidden if anything in the finally throws.

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs (hunk 5 above)
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs]: Was to avoid collisions with the other SocketLoggerExtensions classes in the other packages

File: src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/LongPollingTransport.cs]: it is fine then - as long as we are not losing this information

File: test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs (hunk 6 above)
Comment (moozzyk) [inline test/Microsoft.AspNetCore.SignalR.Tests/EndToEndTests.cs]: Let's leave it for now and we will see if it is bad. This might be helpful when investigating flaky tests in the CI - losing the original exception will make it hard to figure out what happened.

File: src/Microsoft.AspNetCore.Sockets.Client.Http/WebSocketsTransport.cs [hunk 9]
```diff
@@ -47,25 +47,26 @@ public WebSocketsTransport(ILoggerFactory loggerFactory)
             }
 
             _application = application;
+            _connectionId = connectionId;
 
             await Connect(url);
+            _logger.StartTransport(_connectionId);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/WebSocketsTransport.cs]: Should't this go before connect?

File: src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs (hunk 5 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Http/Internal/SocketLoggerExtensions.cs]: It should be SocketClientLoggerExtensions then.

File: src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs [hunk 10]
```diff
@@ -96,8 +96,6 @@ public void Scan()
                 return;
             }
 
-            _logger.ScanningConnections();
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: Are we going to log when a connection is disposed as part of this change or this is separate?

Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/ConnectionManager.cs]: I think we already do
